### ISSUE TITLE
Temporarily disabling grad doesn't work if AMP is enabled

### ISSUE THREAD

**Issue by gahdritz** [pytorch/pytorch#65766]
## 🐛 Bug

grad fails to restart after being temporarily disabled under AMP.

## To Reproduce

Here's some minimal reproduction code:

```
import torch
import torch.nn as nn

class RecycledLinear(nn.Module):
    def __init__(self, no_cycles):
        super(RecycledLinear, self).__init__()

        self.no_cycles = no_cycles
        self.linear = nn.Linear(10, 10)

    def forward(self, inp):
        out = inp
        for i in range(self.no_cycles):
            with torch.set_grad_enabled(i == (self.no_cycles - 1)):
                out = self.linear(out)

        return out


no_cycles = 2
x = torch.rand(10, 10, requires_grad=False).cuda()
model = RecycledLinear(no_cycles).cuda()

with torch.cuda.amp.autocast():
    out = model(x)

print(out.requires_grad)
```

Even though grad is enabled during the final iteration of the network, the output emerges with grad disabled. This happens in spite of the fact that, in the final iteration, `torch.grad.is_enabled()` and `self.linear.weight.requires_grad` are both true.

## Expected behavior

I would expect grad to be enabled on the output, and for the operations in the final iteration of the network to have been recorded properly.

## Environment

PyTorch version: 1.9.1+cu111
Is debug build: False
CUDA used to build PyTorch: 11.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.6 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.10.2
Libc version: glibc-2.27

Python version: 3.9.5 (default, Jun  4 2021, 12:28:51)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-4.15.0-58-generic-x86_64-with-glibc2.27
Is CUDA available: True
CUDA runtime version: 11.1.74
GPU models and configuration:
GPU 0: GeForce RTX 2080 Ti
GPU 1: GeForce RTX 2080 Ti
GPU 2: GeForce RTX 2080 Ti
GPU 3: GeForce RTX 2080 Ti
GPU 4: GeForce RTX 2080 Ti
GPU 5: GeForce RTX 2080 Ti
GPU 6: GeForce RTX 2080 Ti
GPU 7: GeForce RTX 2080 Ti

Nvidia driver version: 455.23.05
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.2
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] light-the-torch==0.3.5
[pip3] numpy==1.19.3
[pip3] pytorch-lightning==1.4.8
[pip3] torch==1.9.1+cu111
[pip3] torchmetrics==0.5.1
[conda] cudatoolkit               11.2.2               he111cf0_8    conda-forge
[conda] light-the-torch           0.3.5                    pypi_0    pypi
[conda] numpy                     1.19.3                   pypi_0    pypi
[conda] pytorch-lightning         1.4.8                    pypi_0    pypi
[conda] torch                     1.9.1+cu111              pypi_0    pypi
[conda] torchmetrics              0.5.1                    pypi_0    pypi

## Additional comments

Any tips for how to work around this issue would be greatly appreciated.

**Comment by albanD** [comment]
I feel like this is just because the first evaluation of your linear layer within amp is in no grad mode. So the temporary weight doesn't require gradients. And it is never re-computed in a differentiable way.
You can make sure that the first forward runs with grad enabled (even if you discard the result) to fix this.

**Comment by gahdritz** [comment]
Correct me if I'm wrong, but regardless of the requires_grad state of the input to the final iteration, the weights in that iteration should still require grad, meaning that the output should also require grad, right? Also, is it not a concern that the behavior is different when AMP is disabled, even though the requires_grad state of all tensors is identical in that case?

As for the fix: insofar as the output of the first layer is the input to the second layer, I don't see how I could discard the output of the former. Or do you mean that I should detach the output of the first iteration before feeding it to the second? If so, the requires_grad property of all tensors in question would be the same as in the reproduction code, unless I'm missing something.

Thanks!

**Comment by albanD** [comment]
I mean that you can add a `self.linear(inp)` before entering the loop and that fixes the issue.

And yes the params do require gradients, but amp is doing some patching for you to get lower precision computation. And the part that goes from the weight to the lower precision weight is cached. So if the first time it runs, it is in no_grad mode, then the cached lower precision weights don't require gradients (even though the full precision ones still do).
The "solution" above is to make sure the caching happens in grad mode.

**Comment by gahdritz** [comment]
Ah, I see now. Thanks for clarifying.

Is this intentional behavior? It seems to me as if it would be more intuitive if the cached FP16 copies of the parameters always inherited the requires_grad property of the originals.

**Comment by albanD** [comment]
We will need to wait for the amp people to answer that. But my guess is that this is a bug.

**Comment by ptrblck** [comment]
Thanks for reporting this issue @gahdritz!
The use case seems to hit a similar limitation as discussed [here](https://discuss.pytorch.org/t/autocast-and-torch-no-grad-unexpected-behaviour/93475/3), which also contains a suggested workaround.
However, it seems that "splitting" the forward pass isn't trivial in your use case based on the usage in your `forward` function. After a quick discussion with @mcarilli a potential workaround with be to clear the autocast cache in each iteration, which re-enabled Autograd via:
```python
class RecycledLinear(nn.Module):
    def __init__(self, no_cycles):
        super(RecycledLinear, self).__init__()

        self.no_cycles = no_cycles
        self.linear = nn.Linear(10, 10)

    def forward(self, inp):
        out = inp
        # Run the model no_cycles times, enabling grad only on the last iteration
        for i in range(self.no_cycles):
            with torch.set_grad_enabled(i == (self.no_cycles - 1)):
                if torch.is_grad_enabled():
                    torch.clear_autocast_cache()
                out = self.linear(out)

        return out
```

**Comment by nicolas-dufour** [comment]
Hey, the bug still exists in pytorch 2.1, quite annoying to fix when using frameworks like lightning

**Comment by ptrblck** [comment]
@nicolas-dufour You can also disable the cache as a workaround if it's blocking you and the suggested approach doesn't work for you.

**Comment by ezyang** [comment]
The fix is actually really really easy. PR coming

**Comment by nicolas-dufour** [comment]
Hey news on this? The bug is still there and makes training slower.
the only workaround i found is to compute the grads and manually zero them

**Comment by ezyang** [comment]
You could try my PR but there is some problem with transformer encoder.

**Comment by JonasGeiping** [comment]
Would it be possible to bump this up into a high-priority issue, or note the problem in the autocast docs? @ezyang 

Here is another minimal repro
```
import torch
import torch.nn as nn
from torch.amp import autocast


class MinimalRepro(nn.Module):
    def __init__(self, dim=64):
        super().__init__()
        self.norm = nn.LayerNorm(dim)
        self.linear = nn.Linear(dim, dim, bias=False)

    def forward(self, x):
        h = x.clone().detach()
        with torch.no_grad():
            # One step without grads
            h = self.linear(self.norm(h)) + x

        # One step with grads
        # torch.clear_autocast_cache() # uncomment this to fix
        h = self.linear(self.norm(h)) + x
        return h


model = MinimalRepro().cuda()
x = torch.randn(2, 64).cuda().requires_grad_()

# First step
with autocast(device_type="cuda", dtype=torch.bfloat16, enabled=True):
    out = model(x)
loss = out.mean()
loss.backward()

print("backward:")
print(f"x grad? {x.grad is not None}")
print(f"norm.weight grad? {model.norm.weight.grad is not None}")
print(f"linear.weight grad? {model.linear.weight.grad is not None}")
```


It was pretty hard (for me) to debug this problem and find my way here, and the solution of `torch.clear_autocast_cache()` is so straightforward that I'm glad I did end up finding it here, but it would be good to prevent future seekers from having to do this. 

What's especially confusing about the issue is that the norm parameters (as in the example) do correctly receive grads, and only the matrix weights are affected, which makes sense after understanding the cause, but is super weird.

Further, using `model = torch.compile(MinimalRepro()).cuda()` actually also fixes the problem, which is another confusing interaction (although I'm glad that it does because this at least was why my prod runs were unaffected).

It's probably my top-1 silent incorrectness bug this year, and reading these issues, it seems to have been around ever since autocast caching was introduced.

**Comment by malfet** [comment]
Per @ezyang , it's likely not hard to fix, so please do not hesitate to propose a PR

**Comment by ezyang** [comment]
No, the problem was the fix PR was easy but it was broken. I did not know (and still do not know how) to fix it

**Comment by damodharg6** [comment]
> Hi, I'd like to investigate this issue and work on a fix. I'll first reproduce the behavior on the latest PyTorch version and examine the autocast cache handling around grad mode transitions.
